internal/watcher/instance_watcher_service.go

```diff
@@ -0,0 +1,133 @@
+// Copyright (c) F5, Inc.
+//
+// This source code is licensed under the Apache License, Version 2.0 license found in the
+// LICENSE file in the root directory of this source tree.
+
+package watcher
+
+import (
+	"context"
+	"log/slog"
+	"time"
+
+	"github.com/nginx/agent/v3/api/grpc/mpi/v1"
+	"github.com/nginx/agent/v3/internal/config"
+	"github.com/nginx/agent/v3/internal/logger"
+	"github.com/nginx/agent/v3/internal/model"
+)
+
+//go:generate go run github.com/maxbrunsfeld/counterfeiter/v6@v6.8.1 -generate
+//counterfeiter:generate . processOperator
+
+//go:generate go run github.com/maxbrunsfeld/counterfeiter/v6@v6.8.1 -generate
+//counterfeiter:generate . processParser
+
+type (
+	processOperator interface {
+		Processes(ctx context.Context) ([]*model.Process, error)
+	}
+
+	processParser interface {
+		Parse(ctx context.Context, processes []*model.Process) []*v1.Instance
+	}
+
+	InstanceWatcherService struct {
+		agentConfig     *config.Config
+		processOperator processOperator
+		processParsers  []processParser
+		cache           []*v1.Instance
+	}
+
+	InstanceUpdates struct {
+		newInstances     []*v1.Instance
+		deletedInstances []*v1.Instance
+	}
+)
+
+func NewInstanceWatcherService(agentConfig *config.Config) *InstanceWatcherService {
+	return &InstanceWatcherService{
+		agentConfig:     agentConfig,
+		processOperator: NewProcessOperator(),
+		processParsers: []processParser{
+			NewNginxProcessParser(),
+		},
+		cache: []*v1.Instance{},
+	}
+}
+
+func (iw *InstanceWatcherService) Watch(ctx context.Context, ch chan<- InstanceUpdates) {
+	monitoringFrequency := iw.agentConfig.Watchers.InstanceWatcher.MonitoringFrequency
+	slog.DebugContext(ctx, "Starting instance watcher monitoring", "monitoring_frequency", monitoringFrequency)
+
+	instanceWatcherTicker := time.NewTicker(monitoringFrequency)
+	defer instanceWatcherTicker.Stop()
+
+	for {
+		select {
+		case <-ctx.Done():
+			close(ch)
+			return
+		case <-instanceWatcherTicker.C:
+			newCtx := context.WithValue(ctx, logger.CorrelationIDContextKey, logger.GenerateCorrelationID())
+			instanceUpdates, err := iw.updates(newCtx)
+			if err != nil {
+				slog.ErrorContext(newCtx, "Instance watcher updates", "error", err)
+			} else {
+				ch <- instanceUpdates
```

Comment:
Maybe we shouldn't send out updates to the channel if there is nothing to update?

Reply:
yes thats a good point let me update that.